global: processors configuration

jmartinm opened the pull request:
* Adds new config variable RECORD_PROCESSORS that allows to
  specify what processors to use depending on the input type.

Signed-off-by: Javier Martin Montull <javier.martin.montull@cern.ch>

jmartinm commented:
Just to get the conversation started, in order to have our own 'data model' we will need a custom marcxml converter. This PR adds the configuration to allow that.

Comments welcome.

egabancho commented on invenio_records/config.py:
```diff
@@ -30,6 +28,12 @@
 
+RECORD_PROCESSORS = {
+    'json': 'json.load',
+    'marcxml': 'invenio_records.manage:convert_marcxml',
```
Could we just put there the function directly?
```python
import json
from .manage import convert_marcxml

RECORD_PROCESSORS = {
    'json': json.load,
    'marcxml': convert_marcxml,
}
```

kaplun replied:
wouldn't it risk circular import errors?

egabancho replied:
Forget about what I said :wink:

jirikuncar commented on invenio_records/manage.py:
```diff
@@ -48,11 +51,10 @@ def convert_marcxml(source):
-    data = processors[input_type](source)
+
+    data = import_string(
+        current_app.config["RECORD_PROCESSORS"][input_type]
```
I would check if it's callable or text type. Then you can go with both options mentioned in https://github.com/inveniosoftware/invenio-records/pull/3/files#r34031365.

jirikuncar commented:
Alternatively to the application configuration variable we may discuss using *entry points* registry (cc @lnielsen).

jmartinm commented on invenio_records/manage.py:
```diff
@@ -48,11 +52,14 @@ def convert_marcxml(source):
+            current_app.config["RECORD_PROCESSORS"][input_type]
+        )(source)
+    else:
+        data = current_app.config["RECORD_PROCESSORS"][input_type](source)
```
Added the check for string.

Do we need the check if ``is callable()`` as well? Do we raise if it is not string nor callable?

jirikuncar replied:
WDYT about this version for better readability?
```python
converter = current_app.config['RECORD_PROCESSORS'][input_type]
if  isinstance(converter, six.string_types):
    converter = import_string(converter)
data = converter(source)
```

jirikuncar commented on invenio_records/manage.py:
```diff
@@ -48,11 +52,14 @@ def convert_marcxml(source):
-    }
-    data = processors[input_type](source)
+
+    if isinstance(current_app.config["RECORD_PROCESSORS"][input_type],
```
can you use single quote for consistency (`current_app.config['RECORD_PROCESSORS']`)?

jmartinm commented:
Thanks for the suggestion @jirikuncar I have amended the code and added a test case.

Can you have a look at the test case to see if it looks ok? ( at least it passes ;) )

PS. Added `Reviewed-by`

jmartinm commented:
PR updated

* Smaller sample xml file
* Better assert for tests
* Use pkg_resources to find sample xml from tests

jirikuncar commented on tests/test_processors.py:
```diff
@@ -0,0 +1,60 @@
+            'marcxml': 'invenio_records.manage:convert_marcxml'
+        }
+        create(open(pkg_resources.resource_filename(
+            "invenio_records", "../tests/data/sample-records.xml"
```
`os.path.join`

jirikuncar commented on tests/test_processors.py:
```diff
@@ -0,0 +1,60 @@
+        self.assertEquals(Record.create.call_count, 2)
+
+
+TEST_SUITE = make_test_suite(TestProcessor)
```
it's not necessary as we are using py.test for running tests in new packages.

jirikuncar commented on tests/test_processors.py:
```diff
@@ -0,0 +1,60 @@
+        self.app.config['RECORD_PROCESSORS'] = {
+            'marcxml': 'invenio_records.manage:convert_marcxml'
+        }
+        create(open(pkg_resources.resource_filename(
```
IMHO we should check location of this file (`__file__`) instead of `pkg_resources.resource_filename('invenio_records')`. If you install it in non-development mode, then the package will be somewhere in *site-packages*.

jmartinm commented:
Thanks for the hints @jirikuncar . Updated.

jirikuncar commented:
Can you add new label to commit message `* NEW Adds new ...`? It looks good now :+1:

tiborsimko commented:
... and also fix "what processors" -> "which processors".

jmartinm commented:
Done.
